main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -538,15 +538,24 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
       }
 
       case ResolvedAst.Expression.Unary(sop, exp, tvar, loc) => sop match {
-        // TODO: Add more
-        case SemanticOperator.Int32Op.Neg =>
+        case SemanticOperator.Float32Op.Neg | SemanticOperator.Float64Op.Neg
+             | SemanticOperator.Int8Op.Neg | SemanticOperator.Int16Op.Neg | SemanticOperator.Int32Op.Neg |SemanticOperator.Int64Op.Neg
+             | SemanticOperator.BigIntOp.Neg =>
           for {
             (constrs, tpe, eff) <- visitExp(exp)
-            resultTyp <- unifyTypeM(tvar, tpe, Type.Int32, loc)
+            resultTyp <- unifyTypeM(tvar, tpe, loc)
+            resultEff = eff
+          } yield (constrs, resultTyp, resultEff)
+
+        case SemanticOperator.Int8Op.Not | SemanticOperator.Int16Op.Not | SemanticOperator.Int32Op.Not |SemanticOperator.Int64Op.Not
+             | SemanticOperator.BigIntOp.Not =>
+          for {
+            (constrs, tpe, eff) <- visitExp(exp)
+            resultTyp <- unifyTypeM(tvar, tpe, loc)
```

Comment:
This does not enforce any type.

You will have to split it into cases and then enforce that it has the expected type.